openapi/json-schema-generic/src/main/scala/endpoints/documented/generic/JsonSchemas.scala

```diff
@@ -0,0 +1,163 @@
+      labelHead: Witness.Aux[L],
+      jsonSchemaHead: JsonSchema[H],
+      jsonSchemaTail: GenericRecord[T]
+    ): GenericRecord[FieldType[L, H] :: T] =
+      new GenericRecord[FieldType[L, H] :: T] {
+        def jsonSchema: Record[FieldType[L, H] :: T] =
+          (field(labelHead.value.name)(jsonSchemaHead) zip jsonSchemaTail.jsonSchema)
+            .invmap[FieldType[L, H] :: T] { case (h, t) => shapelessField[L](h) :: t }(ht => (ht.head, ht.tail))
+      }
+
+    implicit def consCoproduct[L <: Symbol, H, T <: Coproduct](implicit
+      labelHead: Witness.Aux[L],
+      recordHead: GenericRecord[H],
+      taggedTail: GenericTagged[T]
+    ): GenericTagged[FieldType[L, H] :+: T] =
+      new GenericTagged[FieldType[L, H] :+: T] {
+        def jsonSchema: Tagged[FieldType[L, H] :+: T] = {
+          val taggedHead = recordHead.jsonSchema.tagged(labelHead.value.name)
+          taggedHead.orElse(taggedTail.jsonSchema).invmap[FieldType[L, H] :+: T] {
+            case Left(h)  => Inl(shapelessField[L](h))
+            case Right(t) => Inr(t)
+          } {
+            case Inl(h) => Left(h)
+            case Inr(t) => Right(t)
+          }
+        }
+      }
+
+  }
+
+  trait GenericJsonSchemaLowLowPriority {
+
+    trait GenericRecord[A] extends GenericJsonSchema[A] {
+      def jsonSchema: Record[A]
+    }
+
+    trait GenericTagged[A] extends GenericJsonSchema[A] {
+      def jsonSchema: Tagged[A]
+    }
+
+    implicit def recordGeneric[A, R](implicit
+      gen: LabelledGeneric.Aux[A, R],
+      record: GenericRecord[R]
+    ): GenericRecord[A] =
+      new GenericRecord[A] {
+        def jsonSchema: Record[A] = record.jsonSchema.invmap[A](gen.from)(gen.to)
+      }
+
+    implicit def taggedGeneric[A, R](implicit
+      gen: LabelledGeneric.Aux[A, R],
+      tagged: GenericTagged[R]
+    ): GenericTagged[A] =
+      new GenericTagged[A] {
+        def jsonSchema: Tagged[A] = tagged.jsonSchema.invmap[A](gen.from)(gen.to)
+      }
+  }
+
+  /** @return a `JsonSchema[A]` obtained from an implicitly derived `GenericJsonSchema[A]` */
+  def genericJsonSchema[A](implicit genJsonSchema: GenericJsonSchema[A]): JsonSchema[A] = genJsonSchema.jsonSchema
+
+  final class RecordGenericOps[L <: HList](record: Record[L]) {
+
+    def :*: [H](recordHead: Record[H]): RecordGenericOps[H :: L] =
+      new RecordGenericOps(
+        (recordHead zip record).invmap { case (h, l) => h :: l }(hl => (hl.head, hl.tail))
+      )
+
+    def :×: [H](recordHead: Record[H]): RecordGenericOps[H :: L] = recordHead :*: this
+
+    def as[A](implicit gen: Generic.Aux[A, L]): Record[A] = record.invmap(gen.from)(gen.to)
+
+    def tupled[T](implicit
+      tupler: Tupler.Aux[L, T],
+      gen: Generic.Aux[T, L]
+    ): Record[T] = as[T]
+
+  }
+
+  implicit def toRecordGenericOps[A](record: Record[A]): RecordGenericOps[A :: HNil] =
+    new RecordGenericOps[A :: HNil](record.invmap(_ :: HNil)(_.head))
+
+}
```

Comment:
That a bit above my paygrade, but do I understand correctly that by joining `generic.JsonSchemas` and `circe.JsonSchemas` we basically emulate/reimplement `circe-generic`?

Reply:
Exactly! `generic.JsonSchemas` can derives a “schema” for a given algebraic data type and `circe.JsonSchemas` provides circe `Encoder` and `Decoder` for such schemas.